cli/flags.rs

```diff
@@ -51,129 +39,217 @@ Environment variables:
+        .help("Allow environment access"),
+    ).arg(
+      Arg::with_name("allow-run")
+        .long("allow-run")
+        .help("Allow running subprocesses"),
+    ).arg(
+      Arg::with_name("allow-all")
+        .short("A")
+        .long("allow-all")
+        .help("Allow all permissions"),
+    ).arg(
+      Arg::with_name("no-prompt")
+        .long("no-prompt")
+        .help("Do not use prompts"),
+    ).arg(
+      Arg::with_name("log-debug")
+        .short("D")
+        .long("log-debug")
+        .help("Log debug output"),
+    ).arg(
+      Arg::with_name("reload")
+        .short("r")
+        .long("reload")
+        .help("Reload source code cache (recompile TypeScript)"),
+    ).arg(
+      Arg::with_name("v8-options")
+        .long("v8-options")
+        .help("Print V8 command line options"),
+    ).arg(
+      Arg::with_name("v8-flags")
+        .long("v8-flags")
+        .takes_value(true)
+        .require_equals(true)
+        .help("Set V8 command line options"),
+    ).arg(
+      Arg::with_name("types")
+        .long("types")
+        .help("Print runtime TypeScript declarations"),
+    ).arg(
+      Arg::with_name("prefetch")
+        .long("prefetch")
+        .help("Prefetch the dependencies"),
+    ).subcommand(
+      // TODO(bartlomieju): version is not handled properly
+      SubCommand::with_name("info")
+        .about("Show source file related info")
+        .arg(Arg::with_name("file").takes_value(true).required(true)),
+    ).subcommand(
+      // TODO(bartlomieju): version is not handled properly
+      SubCommand::with_name("fmt").about("Format files").arg(
+        Arg::with_name("files")
+          .takes_value(true)
+          .multiple(true)
+          .required(true),
+      ),
+    ).subcommand(
+      // this is a fake subcommand - it's used in conjunction with
+      // AppSettings:AllowExternalSubcommand to treat it as an
+      // entry point script
+      SubCommand::with_name("<script>").about("Script to run"),
+    );
 
-  let rest =
-    set_recognized_flags(&opts, &mut flags, args).map_err(|e| e.to_string())?;
-  Ok((flags, rest, get_usage(&opts)))
+  let matches = clap_app.get_matches_from(args);
+
+  // TODO(bartomieju): compatibility with old "opts" approach - to be refactored
+  let mut rest: Vec<String> = vec![String::from("deno")];
+
+  match matches.subcommand() {
+    ("info", Some(info_match)) => {
+      // TODO(bartlomieju): it still relies on `is_present("info")` check
+      // in `set_recognized_flags`
+      let file: &str = info_match.value_of("file").unwrap();
+      rest.extend(vec![file.to_string()]);
+    }
+    ("fmt", Some(fmt_match)) => {
+      // TODO(bartlomieju): it still relies on `is_present("fmt")` check
+      // in `set_recognized_flags`
+      let files: Vec<String> = fmt_match
+        .values_of("files")
+        .unwrap()
+        .map(String::from)
+        .collect();
+      rest.extend(files);
+    }
+    (script, Some(script_match)) => {
```

Comment:
Could this somehow match only if endswith .ts or .js.

At the moment, typos aren't caught correctly e.g. --allow-ne

Reply:
it behaves exactly like previous version, we can change it further. I'll put up another PR tonight with more refactors.

Reply:
"it behaves exactly like previous version" This isn't quite true (though I think it's a big improvement!) previously --allow-ne would be silently passed through to the script as an arg, now it's treated as a potential script (correctly thowing an error, but a confusing one).

Reply:
@hayd can you paste an example?

Reply:
`deno --allow-ne foo.ts`

Reply:
Huh! Now I try this again, this isn't the case! My mistake! 👍

Reply:
This (new behavior) is _really_ nice:
```
error: Found argument '--allow-ne' which wasn't expected, or isn't valid in this context
	Did you mean --allow-net?

USAGE:
    deno --allow-net

For more information try --help
```